File: data structures/c/stack/stack.c
#include <stdlib.h>

typedef struct {
  size_t *data;
  size_t top;
  size_t maxSize;
} Stack;

Stack createStack(size_t maxSize) {
  Stack s;
  s.maxSize = maxSize;
  s.data = malloc(sizeof(size_t) * s.maxSize);
  s.top = 0;
  return s;
}

int pop(Stack *s) {
  // if stack empty
  if (!s || s->top == 0) {
    return -1;
  }
  int top = s->data[s->top - 1];
  s->data[s->top] = 0;
  s->top--;
  return top;
}

void push(Stack *s, int element) {
  if (!s) {
    return;
  }

  if (s->top == s->maxSize) {
    return;
  }

  s->data[s->top] = element;
  s->top++;
}

int top(Stack *s) {
  if (!s) {
    return -1;
  }

  return s->data[s->top];
}

int size(Stack *s) {
  return s->top;
}

int isEmpty(Stack *s) {
  if (!s || !s->top) {
    return 0;
  }
  return 1;
}

File: data structures/python/stack/stack.py
class Stack:
    def __init__(self, capacity):
        self.data = [None] * capacity
        self.capacity = capacity
        self.current_size = 0

    def push(self, element):
        if self.is_full():
            for i in range(1, self.capacity):
                self.data[i - 1] = self.data[i]
            self.data[self.capacity - 1] = element
        else:
            self.data[self.current_size] = element
            self.current_size += 1

    def pop(self):
        if self.is_empty():
            raise IndexError("Stack empty")
        self.current_size -= 1
        return self.data[self.current_size]

    def size(self):
        return self.current_size

    def is_empty(self):
        return self.current_size == 0

    def is_full(self):
        return self.current_size == self.capacity
